device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/twin/TwinCollection.java

```diff
@@ -0,0 +1,489 @@
+ *             "$lastUpdatedVersion":4,
+ *         },
+ *         "MaxSpeed":{
+ *             "$lastUpdated":"2017-09-21T02:07:44.238Z",
+ *             "$lastUpdatedVersion":4,
+ *             "Value":{
+ *                 "$lastUpdated":"2017-09-21T02:07:44.238Z",
+ *                 "$lastUpdatedVersion":4
+ *             },
+ *             "NewValue":{
+ *                 "$lastUpdated":"2017-09-21T02:07:44.238Z",
+ *                 "$lastUpdatedVersion":4
+ *             }
+ *         }
+ *     },
+ *     "$version":4
+ * }
+ * }
+ * </pre>
+ *
+ * <p> This class exposes the Twin collection with or without metadata as a Map here
+ * user can get both the value and the metadata. For instance, in the above TwinCollection,
+ * {@link #get(Object)} for <b>Color</b> will return <b>White</b> and the {@link #getTwinMetadata(String)}
+ * for <b>Color</b> will return the Object TwinMetadata that contain {@link TwinMetadata#getLastUpdated()}
+ * that will returns the {@code Date} for example <b>2017-09-21T02:07:44.238Z</b>, {@link TwinMetadata#getLastUpdatedBy()}
+ * that will return the {@code String} for example <b>testConfig</b>, {@link TwinMetadata#getLastUpdatedByDigest()}
+ * that will return the {@code String} for example <b>637570515479675333</b>, and {@link TwinMetadata#getLastUpdatedVersion()}
+ * that will return the {@code Integer} for example <b>4</b>.
+ *
+ * <p> For the nested TwinCollection, you can do the same, for instance, the following code will return the
+ * value and metadata of the <b>NewValue</b> nested in <b>MaxSpeed</b>:
+ * <pre>
+ * {@code
+ *      // Get the value of the MaxSpeed, which is a inner TwinCollection.
+ *      TwinCollection innerMaxSpeed = (TwinCollection) twinCollection.get("MaxSpeed");
+ *
+ *      // From the inner TwinCollection, get the value of the NewValue.
+ *      Long maxSpeedNewValue = innerMaxSpeed.get("NewValue");
+ *
+ *      // As in the root TwinCollection, the inner TwinCollection contain its own metadata.
+ *      // So, get the metadata information for the inner NewValue.
+ *      TwinMetadata maxSpeedNewValueMetadata = innerMaxSpeed.getTwinMetadata("NewValue");
+ *      Date newValueLastUpdated = maxSpeedNewValueMetadata.getLastUpdated(); //Shall contain `2017-09-21T02:07:44.238Z`
+ *      Integer newValueLastUpdatedVersion = maxSpeedNewValueMetadata.getLastUpdatedVersion(); //Shall contain `4`
+ * }
+ * </pre>
+ *
+ * @see <a href="https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-device-twins">Understand and use device twins in IoT Hub</a>
+ * @see <a href="https://docs.microsoft.com/en-us/rest/api/iothub/devicetwinapi">Device Twin Api</a>
+ */
+// Unchecked casts of Maps to Map<String, Object> are safe as long as service is returning valid twin json payloads. Since all json keys are Strings, all maps must be Map<String, Object>
+@SuppressWarnings("unchecked")
+public class TwinCollection extends HashMap<String, Object> {
```

Comment:
If this wasn't a HashMap, would that resolve our duplicated methods like `putAll` and `putAllFinal`?

Reply:
Yep. I'll make this class no longer extend HashMap